crates/fluvio-protocol/src/core/bytebuf.rs

```diff
@@ -0,0 +1,301 @@
+use std::io::Error;
+use std::io::ErrorKind;
+
+use bytes::buf::UninitSlice;
+use bytes::{Buf, Bytes, BufMut};
+
+use crate::{Encoder, Decoder, Version};
+
+/// Represnts a SmartModule WASM File bytes.
+///
+/// Provides a `Encoder` and `Decoder` implementation optimized for WASM files
+/// used in SmartModules.
+#[derive(Clone, Debug, Default, PartialEq, Eq)]
+pub struct ByteBuf {
+    inner: Vec<u8>,
+}
+
+impl ByteBuf {
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    pub fn inner(&self) -> Vec<u8> {
+        self.inner.clone()
+    }
+}
+
+impl From<Vec<u8>> for ByteBuf {
+    fn from(bytes: Vec<u8>) -> Self {
+        ByteBuf { inner: bytes }
+    }
+}
+
+impl Decoder for ByteBuf {
+    fn decode<T>(&mut self, src: &mut T, version: Version) -> Result<(), Error>
+    where
+        T: Buf,
+    {
+        let mut len: i32 = 0;
```

Comment:
using i32 here and u32 in encoder.

Reply:
Gotcha! Thanks!

Reply:
Resolved on: 9644e0f by sticking to `u32`